src/api.rs

```diff
@@ -0,0 +1,260 @@
+                transaction_id: transaction_id.clone(),
+                commit_digest: commit_digest.clone(),
+            }),
+            ..Default::default()
+        };
+
+        debug!("request: commit_transaction {:?}", request);
+        let response = self.send_command(request).await?;
+
+        let committed = response
+            .commit_transaction
+            .ok_or(QldbError::UnexpectedResponse(
+                "CommitTransaction requests should return CommitTransaction responses".into(),
+            ))?;
+
+        let remote_transaction_id =
+            committed
+                .transaction_id
+                .ok_or(QldbError::UnexpectedResponse(
+                    "CommitTransaction should always return a transaction_id".into(),
+                ))?;
+
+        if transaction_id != remote_transaction_id {
+            return Err(QldbError::UnexpectedResponse(format!(
+                r#"The committed transaction id did not match our transaction id, this should never happen.
+ The transaction we we committed was {:?}, the server responded with {:?}."#,
+                transaction_id, remote_transaction_id
+            )));
+        }
+
+        let remote_commit_digest = committed
+            .commit_digest
+            .ok_or(QldbError::UnexpectedResponse(
+                "CommitTransaction should always return a commit_digest".into(),
+            ))?;
+
+        if commit_digest != remote_commit_digest {
+            return Err(QldbError::UnexpectedResponse(format!(
+                r#"The commit digest we sent did not match the response sent by the server, this should never happen.
+ The digest we sent was {:?}, the server responded with {:?}."#,
+                commit_digest, remote_commit_digest
+            )));
+        }
+
+        Ok(())
+    }
+
+    async fn end_session(&self, session_token: SessionToken) -> Result<(), QldbError> {
+        let request = SendCommandRequest {
+            session_token: Some(session_token.clone()),
+            end_session: Some(EndSessionRequest {}),
+            ..Default::default()
+        };
+
+        debug!("request: end_session {:?}", request);
+        let response = self.send_command(request).await?;
+
+        response.end_session.ok_or(QldbError::UnexpectedResponse(
+            "EndSession requests should return EndSession responses".into(),
+        ))?;
+
+        Ok(())
+    }
+
+    // FIXME: params, result
+    async fn execute_statement(
+        &self,
+        session_token: &SessionToken,
+        transaction_id: &TransactionId,
+        statement: String,
+    ) -> Result<Option<Page>, QldbError> {
+        let request = SendCommandRequest {
+            session_token: Some(session_token.clone()),
+            execute_statement: Some(ExecuteStatementRequest {
+                transaction_id: transaction_id.clone(),
+                statement: statement,
+                parameters: None,
```

Comment:
Are you going to add parameters here later?

Reply:
Yeah. The main thing is we need Ion hash support before they're actually usable. So, for now, you just can't pass any params in.